### DISCUSSION TITLE
Filters for relationship select in item view

### DISCUSSION THREAD

**Discussion by tv42** [keystonejs/keystone#7424] [Feature requests]
I'd like to limit what items a relationship may pick, in the UI, based on a bool field. For the sake of conversation, consider picking only `Users` with `isAdmin==true`.

I can prevent the bad values in a hook, but that's not user friendly. I'd like to specifically to hide the invalid choices in the UI.

Is that a reasonable amount of work, has anyone done it?

**Comment by fatihsolhan** [comment]
Yes, it is possible with [custom views](https://github.com/keystonejs/keystone/tree/main/examples/custom-field-view)

TLDR
--
Create these components in your keystone project. ( ideally in the components folder )
[FilteredRelationshipSelect.tsx](https://gist.github.com/fatihsolhan/3a5cd6729363c2e389f6f4ac6f3ced2d)
[FilteredRelationship.tsx](https://gist.github.com/fatihsolhan/4f5403ebdc0e793bcb2484ab13b28da3)

Open up the `FilteredRelationship.tsx` file and replace `CUSTOM_FILTER` in the 24th line with your custom filter.

Use this component in your schema field like this:
```js
user: relationship({
  ref: 'User',
  ui: {
    displayMode: 'select',
    views: require.resolve('../components/FilteredRelationship.tsx'),
  },
})
```

---------

## Let me explain how to create those 2 components

#### First, we need to create the `FilteredRelationshipSelect` component

1. Copy https://github.com/keystonejs/keystone/blob/main/packages/core/src/fields/types/relationship/views/RelationshipSelect.tsx file contents and create a component called `FilteredRelationshipSelect.tsx` in your keystone project.
2. Replace `'../../../../'` part of imports with `'@keystone-6/core/'`
3. add `extraFilters = ''` to line 109 after `extraSelection = ''`
4. add `extraFilters?: string` to line 130 after `extraSelection?: string;`
5. replace line 154 with `const where = useFilter(debouncedSearch, list, extraFilters);`
6. replace line 59 with `function useFilter(search: string, list: ListMeta, extraFilters: string) {`
7. replace line 61 with `let conditions: Record<string, any>[] = [JSON.parse(extraFilters)];`
8. replace line 99 with `export const Field = ({`

#### Next, we need to create `FilteredRelationship` component

1. Replace `'../../../../'` part of imports with `'@keystone-6/core/'`
2. Replace `import { RelationshipSelect } from './RelationshipSelect';` with our new created component `import { Field as FilteredRelationshipSelect } from './FilteredRelationshipSelect';`
3. remove the 23rd line since we will only use the Select component not Cards
4. add your custom filter to line 24 as `const CUSTOM_FILTER = '{ "role": { "type": { "equals": "ADMIN" } } }'`
5. Please make sure that `CUSTOM_FILTER` is a valid JSON string.
6. replace `RelationshipSelect` with `FilteredRelationshipSelect` on lines 166 and 535
7. remove lines between 127 to 156 to remove `cards-view` and `count` views since we don't use them
8. pass your custom filter to `FilteredRelationshipSelect` components as `extraFilters={CUSTOM_FILTER}`

replace `CUSTOM_FILTER` in the 24th line with your custom filter.

Use this component in your schema field like this:
```js
user: relationship({
  ref: 'User',
  ui: {
    displayMode: 'select',
    views: require.resolve('../components/FilteredRelationship.tsx'),
  },
})
```

---
**Note to @raveling,  I would love to create a PR to have this feature natively if this is not gonna create complexity on somewhere I don't know**

**Comment by dcousens** [comment]
Hi @tv42,

This use case is **not** something we have forgotten about; and it is on our radar.

There a number of complicating factors and architectural design decisions we are looking into for this feature insofar as it relates to conditional fields generally; sometimes the conditions/filters are dependent on the item context,  sometimes the conditions/filters are dependent on the targeted relationship and other times some kind of global context (for example, time).

We are considering these factors in ongoing work and you should hopefully expect an update on this this year.
In the interim,  you could look at rolling your own as recommended by @fatihsolhan.

Hope this helps :pray:

**Reply by DiesIrae** [reply]
Hi @dcousens, is there any news on this feature? We have several use cases where we had to do bad UX workarounds because of its absence.

In more general terms, dynamic features in form: hide/show, read-only/write, etc based on other fields
